File: python/examples/rest_quickstart.py
#!/usr/bin/env python3
"""Example: Quickstart with synchronous REST API."""

import os
import sys

# Ensure local package is importable when running directly
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from pia import AuthenticationError, PiaClient, RateLimitError


def main() -> None:
    api_key = os.environ.get("PIA_API_KEY", "wi_live_sample_key")
    client = PiaClient(api_key=api_key, debug=True)

    try:
        print("1. Fetching live market prices snapshot...")
        prices = client.market.get_prices()
        print(f"Total instruments tracked: {prices.total}")

        for item in prices.items[:5]:
            print(f"[{item.symbol}] ${item.price} (bid: {item.bid}, ask: {item.ask})")

        print("\n2. Fetching historical 1-hour candles for XAUUSD...")
        candles = client.market.get_candles("XAUUSD", timeframe="1h", limit=10)
        print(f"Received {candles.count} candles for {candles.symbol}")

        telemetry = client.get_rate_limit_info()
        print(f"\nRemaining Minute Quota: {telemetry.remaining}/{telemetry.limit}")
        print(f"Remaining Daily Quota: {telemetry.daily_remaining}/{telemetry.daily_limit}")

    except AuthenticationError:
        print("Authentication error: Verify that your API key is valid.")
    except RateLimitError as e:
        print(f"Rate limited! Retry after {e.retry_after_seconds}s")
    except Exception as e:
        print(f"Error: {e}")
    finally:
        client.close()


if __name__ == "__main__":
    main()

File: typescript/examples/rest-quickstart.ts
import { PiaClient, RateLimitError } from "../src";

async function main() {
  const client = new PiaClient({
    apiKey: process.env.PIA_API_KEY || "wi_live_sample_key",
    debug: true,
  });

  try {
    // 1. Fetch Multi-Asset Price Snapshot
    console.log("Fetching live market prices...");
    const prices = await client.market.getPrices();
    console.log(`Total instruments: ${prices.total}`);

    for (const item of prices.items.slice(0, 5)) {
      console.log(`[${item.symbol}] $${item.price} (bid: ${item.bid}, ask: ${item.ask})`);
    }

    // 2. Fetch Candlesticks
    console.log("\nFetching historical 1-hour candles for XAUUSD...");
    const candles = await client.market.getCandles("XAUUSD", {
      timeframe: "1h",
      limit: 10,
    });
    console.log(`Received ${candles.count} candles.`);

    // 3. Inspect telemetry
    const telemetry = client.getRateLimitInfo();
    console.log(`\nRemaining Minute Quota: ${telemetry.remaining}/${telemetry.limit}`);
    console.log(`Remaining Daily Quota: ${telemetry.dailyRemaining}/${telemetry.dailyLimit}`);
  } catch (err) {
    if (err instanceof RateLimitError) {
      console.error(`Rate limit exceeded! Retry after ${err.retryAfterSeconds}s`);
    } else {
      console.error("API error:", err);
    }
  }
}

main().catch(console.error);
